datasetinsights/io/tracker/factory.py

```diff
@@ -0,0 +1,115 @@
+import threading
+
+from datasetinsights.constants import TIMESTAMP_SUFFIX
+from datasetinsights.io.tracker.mlflow import MLFlowTracker
+
+
+class TrackerFactory:
+    """Factory: responsible for creating and holding singleton instance
+        of tracker classes"""
+
+    TRACKER = "tracker"
+    CLIENT_ID = "client_id"
+    HOST_ID = "host"
+    EXP_NAME = "experiment"
+    RUN_NAME = "run"
+    MLFLOW_TRACKER = "mlflow"
+    DEFAULT_RUN_NAME = "run_" + TIMESTAMP_SUFFIX
+    __singleton_lock = threading.Lock()
+    __tracker_instance = None
+    RUN_FAILED = "FAILED"
+
+    @staticmethod
+    def create(config, tracker_type):
+        """create tracker class object.
+
+        Args:
+            config : config object, holds server details
+            tracker_type : type of tracker
+
+        Returns:
+            tracker instance.
+        """
+        if TrackerFactory.MLFLOW_TRACKER == tracker_type:
+
+            tracker = config.get(TrackerFactory.TRACKER, None)
+            if tracker and tracker.get(TrackerFactory.MLFLOW_TRACKER, None):
+                mlflow_config = tracker.get(TrackerFactory.MLFLOW_TRACKER)
+                host_id = mlflow_config.get(TrackerFactory.HOST_ID, None)
+                if host_id:
+                    client_id = mlflow_config.get(
+                        TrackerFactory.CLIENT_ID, None
+                    )
+                    exp_name = mlflow_config.get(TrackerFactory.EXP_NAME, None)
+                    TrackerFactory.update_run_name(mlflow_config)
+                    mlflow_tracker = TrackerFactory._mlflow_tracker_instance(
+                        host_id=host_id, client_id=client_id, exp_name=exp_name
+                    ).get_mlflow()
+                    mlflow_tracker.start_run(
+                        run_name=TrackerFactory.DEFAULT_RUN_NAME
+                    )
+                    return mlflow_tracker
+            return TrackerFactory._null_tracker()
+        else:
+            raise NotImplementedError(f"Unknown tracker {tracker_type}!")
```

Comment:
I think ValueError is more appropriate. Even better, create an `InvalidTrackerError` so it's very obvious we're raising the exception as opposed to some downstream dependency